lib/lhm/throttler/slave_lag.rb

```diff
@@ -43,55 +45,78 @@ def throttle_seconds
         end
       end
 
-      def slave_hosts
-        slaves = get_slaves.map { |slave_host| slave_host.partition(':')[0] }
-          .delete_if { |slave| slave == 'localhost' || slave == '127.0.0.1' }
-        Lhm.logger.info "Detected slaves: #{slaves.join(',')}"
-        slaves
+      def slaves
+        @slaves[@connection] ||= get_slaves
       end
 
       def get_slaves
-        @connection.select_values(SQL_SELECT_SLAVE_HOSTS)
+        slaves = []
+        slave_hosts = master_slave_hosts
+        while slave_hosts.any? do
+          host = slave_hosts.pop
+          slave = Slave.new(host, @get_config)
+          if slaves.map(&:host).exclude?(host) && slave.connection
+            slaves << slave
+            slave_hosts.concat(slave.slave_hosts)
```

Comment:
we have a connection to the master, and we need to get the slaves for that master, then connect to each of those slaves, then get those slaves' slaves, etc. on down the chain.

`slaves` is an array of the slave objects that have their replication lag checked, which is what this method actually returns.  `slave_hosts` is an array consisting of all IP addresses slaving off of anything in the chain.  this `while` loop pops off each IP address in the `slave_hosts` array and creates a `Slave` object from that IP address.

so for example, when we get far enough down the chain this will try to connect to a hadoop instance, since that instance is slaving off the reader in a shard.  calling `slave.slave_hosts` on the reader will return the hadoop node's IP, that gets added to the `slave_hosts` array, the `while slave_hosts.any?` loop will eventually try to connect to that instance and fail, so `slave.connection` will be `nil`, this `if` condition will not be met and that slave will not be added to our final array of slaves that we want to check for slave lag.

so TL;DR the `slave_hosts` array exists to keep track of all slaves in the chain, and `slaves` is a separate array of slaves that we actually want to check.